proposals/0008-package-manager-testing.md

```diff
@@ -0,0 +1,194 @@
+            └── Test.swift
+
+Or:
+
+    Package
+    └── Sources
+        ├── Foo.swift
+        └── Tests
+            └── Test.swift
+
+Or, for simpler projects:
+
+    Package
+    ├── Sources
+    │   └── Foo.swift
+    └── Tests
+        └── TestFoo.swift
+        
+The filename: `TestFoo.swift` is arbitrary.
+
+In the examples above
+a test case is created for the module `Foo`
+based on the sources in the relevant subdirectories.
+
+Additionally we will support directories called `FooTests`.
+This layout style is prevalent in existing open source projects
+and supporting it will minimize vexation for their authors.
+However in the interest of consistency and the corresponding 
+reduction of cognitive-load when examining new Swift packages
+we will not recommend this layout. For example:
+
+    Package
+    └── Sources
+    │   └── Foo.swift
+    └── FooTests
+        └── Test.swift
+
+Additionally, we propose that building a module
+also builds that module's corresponding tests.
+Although this would result in slightly increased build times,
+we believe that tests are important enough to justify this
+(one might even consider slow building tests to be a code smell).
+We would prefer to go even further by executing the tests
+each time a module is built as well,
+but we understand that this would impede debug cycles.
+
+As an exception, when packages are built in release mode we will
+not build tests because for release builds we should not enable
+testability. However, considering the need for release-mode testing
+this will be a future direction.
+
+### Command-Line Interface
+
+We propose the following syntax to execute tests:
+
+    $ swift build --test
+
+Or:
+
+    $ swift build -t
+
+In the future, we may choose to promote the `--test` option
+to be a subcommand of the `swift` command itself:
+
+    $ swift test
+
+However, any such decision would warrant extensive design consideration,
+so as to avoid polluting or crowding the command-line interface.
+Should there be sufficient demand and justification for it, though,
+it would be straightforward to add this functionality.
+
+### Command Output
+
+Executing a test from the terminal will produce user-readable output.
+This should incorporate colorization and other formatting
+similar to other testing tools
+to indicate the success and failure of different tests.
+For example:
+
+    $ swift test --output module
+    Running tests for PackageX (x/100)
+    .........x.....x...................
+
+    Completed
+    Elapsed time: 0.2s
+
+    98 Success
+     2 Failure
+     1 Warning
+
+    FAILURE: Tests/TestsA.swift:24 testFoo()
+    XCTAssertTrue expected true, got false
+
+    FAILURE: Tests/TestsB.swift:10 testBar()
+    XCTAssertEqual
+
+    WARNING: Tests/TestsC.swift:1
+    "Some Warning"
+
+An additional option may be passed to the testing command
+to output JUnit-style XML or other formats that can be integrated
+with continuous integration (CI) and other systems.
```

Comment:
Would this system use a plugin architecture to allow users to specify their own test reporters? For example, one that outputs the results in a JSON format? Or would each reporter need to be merged into the swiftpm codebase?